src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/FormattingStructureValidationPass.cs

```diff
@@ -0,0 +1,115 @@
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Formatting
+{
+    internal class FormattingStructureValidationPass : FormattingPassBase
+    {
+        private readonly ILogger _logger;
+
+        public FormattingStructureValidationPass(
+            RazorDocumentMappingService documentMappingService,
+            FilePathNormalizer filePathNormalizer,
+            ILanguageServer server,
+            ILoggerFactory loggerFactory)
+            : base(documentMappingService, filePathNormalizer, server)
+        {
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger<CSharpOnTypeFormattingPass>();
+        }
+
+        // We want this to run at the very end.
+        public override int Order => DefaultOrder + 1000;
+
+        public override FormattingResult Execute(FormattingContext context, FormattingResult result)
+        {
+            Debug.Assert(result.Kind == RazorLanguageKind.Razor, "This method shouldn't be called for projected document edits.");
+
+            if (result.Edits.Length == 0)
+            {
+                // No op.
+                return result;
+            }
+
+            if (FormatsOutsidePureCSharpBlocks(context, result))
+            {
+                return new FormattingResult(Array.Empty<TextEdit>());
+            }
+
+            return result;
+        }
+
+        private bool FormatsOutsidePureCSharpBlocks(FormattingContext context, FormattingResult result)
+        {
+            if (!context.IsFormatOnType)
+            {
+                // We don't care about regular formatting for now.
+                return false;
+            }
+
+            var text = context.SourceText;
+            var changes = result.Edits.Select(e => e.AsTextChange(text));
+            var changedText = text.WithChanges(changes);
+            var affectedSpan = changedText.GetEncompassingTextChangeRange(text).Span;
+            var affectedRange = affectedSpan.AsRange(text);
+
+            var syntaxTree = context.CodeDocument.GetSyntaxTree();
+            var nodes = syntaxTree.GetCodeBlockDirectives();
+
+            var affectedCodeDirective = nodes.FirstOrDefault(n =>
+            {
+                var range = n.GetRange(context.CodeDocument.Source);
+                return range.Contains(affectedRange);
+            });
+
+            if (affectedCodeDirective == null)
+            {
+                _logger.LogDebug("A formatting result was rejected because it was going to format outside of a code block directive.");
+
+                return true;
+            }
+
+            if (!(affectedCodeDirective.Body is RazorDirectiveBodySyntax directiveBody))
+            {
+                // This can't happen realistically. Just being defensive.
+                return false;
```

Comment:
Wont  this  result  in  it attempting  to format the  doc?  Ditto witht he other statements

Reply:
Yeah this is intentional. We already know the range is contained within this code block. The only ever realistic scenario is when the code block is empty in which case it doesn't matter whether we accept or reject the formatting. Decided to let it format in these cases just to see what corner cases I may have missed.